### PR TITLE
#2446 - Revise Offering Length Validations

### PR DESCRIPTION
**Pull request by lewischen-aot**
- Removed the current validation on minimum offering lengths which require FT offerings to be at least 84 days, and PT offerings to be at least 42 days for study breaks validation
  - Added validation on minimum offering weeks `studyPeriodMinWeeks` for FT offerings (12 weeks) and PT offerings (6 weeks)
- Removed any associated warning message banner as well that is displayed when offering length does not meet those requirements.
- Applied a new validation on FT offerings which require that the Total Funded Weeks field is 12 or higher. If 11 or below, display an warning banner message below the "Study Period Breakdown" which reads:
  - "Full-time offerings must be at least 12 weeks of study or longer to be eligible. Any shorter offerings can be submitted but will require SABC review."
- Applied a new validation on PT offerings which require that the Total Funded Weeks field is 6 or higher. If 5 or below, display an warning banner message below the "Study Period Breakdown" which reads:
  - "Part-time offerings must be at least 6 weeks of study or longer to be eligible. Any shorter offerings can be submitted but will require SABC review."
- Adjusted validations of PIR process during student application to remove the current validation on students when they enter study length which does not comply with the number of weeks policy.
- Adjusted the validations imposed on Bulk Offering excel upload, as applicable.
- Created new E2E tests for the new scenarios.
image

Screenshot of warning banner for part-time offering with invalid total funded weeks and no study breaks
image

Screenshot of warning banner for part-time offering with invalid total funded weeks and study breaks
image

Screenshot of warning banner with total funded days more than max allowed days
image

Screenshot of error text for validating PIR program max allowed days during student application
image

Screenshot of warning message during bulk-upload validation for full-time offerings
image

### CODE REVIEW

File: sources/packages/backend/apps/api/src/services/education-program-offering/education-program-offering-validation.models.ts [hunk 1]
```diff
@@ -587,7 +601,7 @@ export class OfferingValidationModel {
   @HasValidOfferingPeriodForFundedDays(
     studyStartDateProperty,
     studyEndDateProperty,
-    studyPeriodMinLength,
+    studyPeriodMinWeeks,
```
Comment (sh16011993) [inline sources/packages/backend/apps/api/src/services/education-program-offering/education-program-offering-validation.models.ts:604]: suggestion: `studyPeriodMinFundedWeeks`

File: sources/packages/backend/apps/api/src/services/education-program-offering/education-program-offering-validation.models.ts [hunk 2]
```diff
@@ -587,7 +601,7 @@ export class OfferingValidationModel {
   @HasValidOfferingPeriodForFundedDays(
     studyStartDateProperty,
     studyEndDateProperty,
-    studyPeriodMinLength,
```
Comment (sh16011993) [inline sources/packages/backend/apps/api/src/services/education-program-offering/education-program-offering-validation.models.ts:590]: Have we removed the code related to this function and the respective validation `studyPeriodMinLength` from the `studyEndDate` property?

File: sources/packages/backend/apps/api/src/services/education-program-offering/custom-validators/has-valid-offering-period-for-funded-days.ts [hunk 3]
```diff
@@ -55,14 +55,14 @@ class HasValidOfferingPeriodForFundedDaysConstraint
  * allowed study period amount of days.
  * @param startPeriodProperty property of the model that identifies the offering start date.
  * @param endPeriodProperty property of the model that identifies the offering end date.
- * @param offeringMinDaysAllowed study period minimum length in number of days.
+ * @param offeringMinWeeksAllowed study period minimum length in number of weeks.
  * @param validationOptions validations options.
  * @returns true if the study period is valid, otherwise, false.
  */
 export function HasValidOfferingPeriodForFundedDays(
```
Comment (sh16011993) [inline sources/packages/backend/apps/api/src/services/education-program-offering/custom-validators/has-valid-offering-period-for-funded-days.ts:62]: `HasValidOfferingPeriodForFundedWeeks` ?

File: sources/packages/backend/apps/api/src/auth/roles.enum.ts [hunk 4]
```diff
@@ -8,6 +8,7 @@ export enum Role {
   AESTManualTriggerReassessment = "aest-manual-trigger-reassessment",
   AESTReports = "aest-reports",
   AESTCreateInstitution = "aest-create-institution",
+  AESTEditCASSupplierInfo = "aest-edit-cas-supplier-info",
```
Comment (sh16011993) [inline sources/packages/backend/apps/api/src/auth/roles.enum.ts:11]: Is this getting used?

Comment (dheepak-aot) [reply sources/packages/backend/apps/api/src/auth/roles.enum.ts:11]: This constant should not be part of the PR. It is part of [Andre PR](https://github.com/bcgov/SIMS/pull/3674)
I know without this E2E tests will fail but that will happen only until his code is merged.

File: sources/packages/backend/apps/api/src/services/education-program-offering/education-program-offering-validation.models.ts [hunk 5]
```diff
@@ -52,6 +52,8 @@ import {
   OFFERING_STUDY_PERIOD_MAX_DAYS,
   OFFERING_STUDY_PERIOD_MIN_DAYS_FULL_TIME,
   OFFERING_STUDY_PERIOD_MIN_DAYS_PART_TIME,
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/services/education-program-offering/education-program-offering-validation.models.ts:54]: The constants `OFFERING_STUDY_PERIOD_MIN_DAYS_FULL_TIME` and `OFFERING_STUDY_PERIOD_MIN_DAYS_PART_TIME` are no longer required and need to be removed from all places.

Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/services/education-program-offering/education-program-offering-validation.models.ts:54]: Also the validator `studyPeriodMinLength` is not required anymore and it needs to be removed unless I am missing something. 

image

File: sources/packages/forms/src/form-definitions/educationprogramoffering.json
Comment (sh16011993) [inline sources/packages/forms/src/form-definitions/educationprogramoffering.json:1]: Is it due to the formatting issue that it is showing everything removed and added again?

File: sources/packages/forms/src/form-definitions/sfaa2023-24.json
Comment (sh16011993) [inline sources/packages/forms/src/form-definitions/sfaa2023-24.json:1]: I can see some old form changes have gotten reverted by mistake. For instance, 15 MB to 4MB for file uploads etc.

File: sources/packages/backend/apps/api/src/services/education-program-offering/custom-validators/has-valid-offering-period-for-funded-days.ts [hunk 6]
```diff
@@ -20,32 +20,32 @@ class HasValidOfferingPeriodForFundedDaysConstraint
...
       calculatedStudyBreaksAndWeeks.fundedStudyPeriodDays <=
         OFFERING_STUDY_PERIOD_MAX_DAYS
     );
   }
 
   defaultMessage(args: ValidationArguments) {
-    const offeringMinDaysAllowedValue = this.getOfferingMinAllowedDays(args);
-    return `The funded study amount of days is ineligible for StudentAid BC funding. Your dates must be between ${offeringMinDaysAllowedValue} to ${OFFERING_STUDY_PERIOD_MAX_DAYS} days.`;
+    const offeringMinWeeksAllowedValue = this.getOfferingMinAllowedWeeks(args);
+    return `The funded study amount of days is ineligible for StudentAid BC funding. Your offering must be at least ${offeringMinWeeksAllowedValue} weeks of study or longer to be eligible.`;
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/services/education-program-offering/custom-validators/has-valid-offering-period-for-funded-days.ts:38]: The message isn't accurate with respect to logic anymore. 
The logic validates if the `offering funded weeks >= offeringMinWeeksAllowedValue` and also `offering funded days is <= 365 days`.  
Hence either the message must be updated to be in sync with the validation 
or
minimum funded weeks must use a separate constraint and message.

File: sources/packages/backend/apps/api/src/services/education-program-offering/custom-validators/has-valid-offering-period-for-funded-days.ts [hunk 7]
```diff
@@ -20,32 +20,32 @@ class HasValidOfferingPeriodForFundedDaysConstraint
   implements ValidatorConstraintInterface
 {
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/services/education-program-offering/custom-validators/has-valid-offering-period-for-funded-days.ts:21]: The constraint name(also the file name) `HasValidOfferingPeriodForFundedDaysConstraint` is not accurate anymore IMO.  The validation is not just about funded days anymore. It is based on funded days and weeks.

**Review by sh16011993 — COMMENTED** [review]
Left a few comments. Please have a look.

File: sources/packages/backend/apps/api/src/services/education-program-offering/custom-validators/has-valid-offering-period-for-funded-days.ts (hunk 6 above)
Comment (dheepak-aot) [reply sources/packages/backend/apps/api/src/services/education-program-offering/custom-validators/has-valid-offering-period-for-funded-days.ts:38]: @andrewsignori-aot @andrepestana-aot @guru-aot @bidyashish @katrina-aot-git 

Do you have any preference about what approach to take here - 

1. Same constraint and a common warning message

OR

2. New constraint and separate warning message. ?

File: sources/packages/forms/src/form-definitions/educationprogramoffering.json [hunk 8]
```diff
@@ -1,5759 +1,5763 @@
...
+              "protected": false,
+              "unique": false,
+              "persistent": false,
+              "clearOnHide": true,
+              "refreshOn": "",
+              "redrawOn": "",
+              "dataGridLabel": false,
+              "labelPosition": "top",
+              "description": "",
+
```
Comment (dheepak-aot) [inline sources/packages/forms/src/form-definitions/educationprogramoffering.json:5758]: Why does it have 365 as default value?

File: sources/packages/forms/src/form-definitions/educationprogramoffering.json [hunk 9]
```diff
@@ -1,5759 +1,5763 @@
...
-      },
-      "properties": {},
-      "lockKey": true,
-      "customDefaultValue": "",
-      "calculateValue": "value = data.offeringIntensity === 'Full Time' ? 84 : 42;"
-    },
-    {
-      "input": true,
-      "tableView": true,
-      "key": "studyPeriodMaxDays",
```
Comment (dheepak-aot) [inline sources/packages/forms/src/form-definitions/educationprogramoffering.json:5742]: Was there a requirement to remove the max days validation ? and hence the `studyPeriodMaxDays` hidden variable?

File: sources/packages/forms/src/form-definitions/educationprogramoffering.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (dheepak-aot) [reply sources/packages/forms/src/form-definitions/educationprogramoffering.json:2021]: Yes that is right according to the AC.

File: sources/packages/forms/src/form-definitions/educationprogramoffering.json [hunk 10]
```diff
@@ -1,5759 +1,5763 @@
...
-                  "labelMargin": "",
-                  "tag": "p",
-                  "className": "alert alert-warning fa fa-exclamation-triangle w-100",
-                  "attrs": [
-                    {
-                      "attr": "",
-                      "value": ""
-                    }
-                  ],
-                  "content": "<strong class=\"ml-2\">This study period is ineligible for StudentAid BC funding</strong>\n<br>\n<span>Your dates must be between {{data.studyPeriodMinDays}} to {{data.studyPeriodMaxDays}} days.</span>",
```
Comment (dheepak-aot) [inline sources/packages/forms/src/form-definitions/educationprogramoffering.json:2021]: Question to business and other devs.

This warning banner currently covers 2 validation and it's respective message. 

1. Minimum study period days  must be (84 for FT and 42 for PT)
2. Maximum study period days must be 365.

As per the current ticket we are updating condition 1. 

1. Minimum funded study weeks  must be (12 for FT and 6 for PT)
2. Maximum study period days must be 365.

Considering that the validation 1 is replaced by minimum study weeks. The warning banner message is now updated. 

But the updated warning banner message is completely ignoring the maximum 365 days validation context in it. 

Hence either we need to update the warning banner message to have a common context or have a separate banner for condition 2.

File: sources/packages/backend/apps/api/src/services/education-program-offering/custom-validators/has-valid-offering-period-for-funded-days.ts (hunk 6 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/services/education-program-offering/custom-validators/has-valid-offering-period-for-funded-days.ts:38]: Option one would be more aligned with the current ticket ACs. The specific message to conciliate both into one sentence should come from business.

The second one can provide more precise warnings but it would also demand business providing separate content messages and a new banner to be displayed in the UI.

For the sake of this PR I would go with option 1.

File: sources/packages/forms/src/form-definitions/sfaa2022-23.json [hunk 11]
```diff
@@ -7993,7 +8012,7 @@
                     {
                       "components": [
                         {
-                          "html": "<p style=\"text-align:right;\">15MB file limit each</p>\n",
+                          "html": "<p style=\"text-align:right;\">4MB file limit each</p>\n",
```
Comment (dheepak-aot) [inline sources/packages/forms/src/form-definitions/sfaa2022-23.json:8015]: The form definitions does not seem to be updated on top of latest code from main. This is reverting the change implemented in #2870

File: sources/packages/backend/apps/api/src/services/education-program-offering/custom-validators/has-valid-offering-period-for-funded-days.ts (hunk 6 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/services/education-program-offering/custom-validators/has-valid-offering-period-for-funded-days.ts:38]: On second thought, the new messages are specific for the minimum validation, right?
If yes, should we not segregate the validations to achieve the ticket ACs?

File: sources/packages/forms/src/form-definitions/sfaa2022-23.json [hunk 12]
```diff
@@ -2514,7 +2514,7 @@
...
+                                "custom": "// To calculate the time difference of two dates\r\nconst Difference_In_Time =\r\n  new Date(data.studyendDate).getTime() -\r\n  new Date(data.studystartDate).getTime();\r\n\r\n// To calculate the no. of days between two dates\r\nconst Difference_In_Days = Difference_In_Time / (1000 * 3600 * 24);\r\n// To calculate the no. of weeks between two dates\r\nconst Difference_In_Weeks = Difference_In_Days / 7;\r\nvalid = data.studystartDate\r\n  ? Difference_In_Time >= 0 \r\n    ? Difference_In_Weeks >= data.studyPeriodMinWeeks && Difference_In_Days <= data.studyPeriodMaxDays\r\n      ? true\r\n      : `Invalid Study Period, ${data.offeringIntensityText} programs must be at least ${data.studyPeriodMinWeeks} weeks and less than ${data.studyPeriodMaxDays} days`\r\n    : \"Study Start date must be before Study End date\"\r\n  : \"Please provide Study Start date\";",
```
Comment (dheepak-aot) [inline sources/packages/forms/src/form-definitions/sfaa2022-23.json:2517]: The requirement in the ticket mention the removal of existing minimum validation in PIR section of student application and not to replace it with weeks. 

image

File: sources/packages/backend/apps/api/src/services/education-program-offering/custom-validators/has-valid-offering-period-for-funded-days.ts (hunk 6 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/services/education-program-offering/custom-validators/has-valid-offering-period-for-funded-days.ts:38]: After the quick sync, the assumptions are below.
- Currently, we have 2 validations for the max 365.
  - The first validates the 365 between the start and end date.
  - The second validates the 365 funded days between the start and end date and also the minimum amount of funded days (to be changed to validate funded weeks).
- Both validators generate the same warning key that results in the same banner in the form.io UI and same message for the bulk upload.

### Proposed solution

- Keep the first decorator validating only the 365 between the start and end date (not the funded values).
- The first validator should display the banner where it is currently, right below the Start and End dates.
  - Business to provide the content message.
- The second validator will check only the minimum funded weeks, as requested in this ticket, and display the messages with the contents as requested in this ticket. The banner will be displayed right below the study breaks following the ticket initial requirements.

File: sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.validateOffering.e2e-spec.ts [hunk 13]
```diff
@@ -55,77 +60,42 @@ describe("EducationProgramOfferingInstitutionsController(e2e)-validateOffering",
...
+      },
     );
-  });
-
-  it("Should validate an offering when passed valid data.", async () => {
-    // Arrange
-    const institutionUserToken = await getInstitutionToken(
-      InstitutionTokenTypes.CollegeFUser,
+    fakeFullTimeEducationProgram.sabcCode = faker.random.alpha({ count: 4 });
+    fakeFullTimeEducationProgram.deliveredOnline = true;
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.validateOffering.e2e-spec.ts:80]: Why is it not added to initial values like `fakePartTimeEducationProgram` as the existing factory supports it.

File: sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.validateOffering.e2e-spec.ts [hunk 14]
```diff
@@ -231,6 +201,440 @@ describe("EducationProgramOfferingInstitutionsController(e2e)-validateOffering",
       });
   });
 
+  it("Should validate an offering when valid data without study breaks is passed in for a full-time education program.", async () => {
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.validateOffering.e2e-spec.ts:204]: Why is the program intensity here an aspect of tests? I did not follow the reason.

File: sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.validateOffering.e2e-spec.ts [hunk 15]
```diff
@@ -231,6 +201,440 @@ describe("EducationProgramOfferingInstitutionsController(e2e)-validateOffering",
...
+        studyPeriodBreakdown: {
+          fundedStudyPeriodDays: 82.6,
+          totalDays: 86,
+          totalFundedWeeks: 12,
+          unfundedStudyPeriodDays: 3.4,
+        },
+      });
+  });
+
+  it("Should return error and warning when data with total funded weeks less than minimum allowed weeks and study breaks is passed in for a full-time education program.", async () => {
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.validateOffering.e2e-spec.ts:356]: info and warning?

File: sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.validateOffering.e2e-spec.ts (hunk 14 above)
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.validateOffering.e2e-spec.ts:204]: The program intensity does not confirm the offering intensity directly. Offering intensity is what provided in the offering payload.

File: sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.validateOffering.e2e-spec.ts (hunk 15 above)
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.validateOffering.e2e-spec.ts:356]: Please add a test for full time without study break as well.

File: sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.validateOffering.e2e-spec.ts [hunk 16]
```diff
@@ -231,6 +201,440 @@ describe("EducationProgramOfferingInstitutionsController(e2e)-validateOffering",
...
+    };
+
+    // Act/Assert
+    await request(app.getHttpServer())
+      .post(endpoint)
+      .send(payload)
+      .auth(institutionUserToken, BEARER_AUTH_TYPE)
+      .expect(HttpStatus.OK)
+      .expect({
+        offeringStatus: "Creation pending",
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.validateOffering.e2e-spec.ts:612]: I would rather use enum instead of string here. `OfferingStatus.CreationPending`

File: sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.validateOffering.e2e-spec.ts [hunk 17]
```diff
@@ -231,6 +201,440 @@ describe("EducationProgramOfferingInstitutionsController(e2e)-validateOffering",
...
+        studyPeriodBreakdown: {
+          fundedStudyPeriodDays: 39.7,
+          totalDays: 47,
+          totalFundedWeeks: 6,
+          unfundedStudyPeriodDays: 7.3,
+        },
+      });
+  });
+
+  it("Should return error and warning when data with total funded weeks less than minimum allowed weeks and study breaks is passed in for a part-time education program.", async () => {
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.validateOffering.e2e-spec.ts:574]: Please add a test for part time without study break as well which would result in 6 minimum funded weeks warning.

**Comment by dheepak-aot** [comment]
Nice work. Please take a look at the comments. 

Please add at least one FT and PT offering test to following test suites. 

`education-program-offering.institutions.controller.createOffering.e2e-spec.ts`

`education-program-offering.institutions.controller.createOffering.e2e-spec.ts`

`education-program-offering.institutions.controller.bulkInsert.e2e-spec.ts`

File: sources/packages/backend/apps/api/src/services/education-program-offering/custom-validators/has-valid-offering-period-for-funded-days.ts [hunk 18]
```diff
@@ -5,78 +5,75 @@ import {
...
 /**
  * For an offering that contains study breaks, execute the calculation
- * of the funded study period to validate if it matches the mix and max
- * allowed study period amount of days.
+ * of the funded study period to validate if it matches the minimum
+ * allowed study period amount of weeks.
  */
 @ValidatorConstraint()
-class HasValidOfferingPeriodForFundedDaysConstraint
+class HasValidOfferingPeriodForFundedWeeksConstraint
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/services/education-program-offering/custom-validators/has-valid-offering-period-for-funded-days.ts:17]: Please rename the file as well.

File: sources/packages/backend/apps/api/src/services/education-program-offering/custom-validators/has-valid-offering-period-for-funded-days.ts [hunk 19]
```diff
@@ -5,78 +5,75 @@ import {
...
+      calculatedStudyBreaksAndWeeks.totalFundedWeeks >=
+      offeringMinWeeksAllowedValue
     );
   }
 
   defaultMessage(args: ValidationArguments) {
-    const offeringMinDaysAllowedValue = this.getOfferingMinAllowedDays(args);
-    return `The funded study amount of days is ineligible for StudentAid BC funding. Your dates must be between ${offeringMinDaysAllowedValue} to ${OFFERING_STUDY_PERIOD_MAX_DAYS} days.`;
+    const offeringMinWeeksAllowedValue = this.getOfferingMinAllowedWeeks(args);
+    return `The funded study amount of days is ineligible for StudentAid BC funding. Your offering must be at least ${offeringMinWeeksAllowedValue} weeks of study or longer to be eligible.`;
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/services/education-program-offering/custom-validators/has-valid-offering-period-for-funded-days.ts:35]: This message is displayed in UI only during bulk offering upload. Please double check with business on the content of this message. As the AC did not state what must be the message for min weeks validation on bulk offerings upload.

File: sources/packages/backend/apps/api/src/auth/roles.enum.ts (hunk 4 above)
Comment (dheepak-aot) [reply sources/packages/backend/apps/api/src/auth/roles.enum.ts:11]: This role is still not removed.

File: sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.bulkInsert.e2e-spec.ts [hunk 20]
```diff
@@ -221,35 +221,38 @@ describe("EducationProgramOfferingInstitutionsController(e2e)-bulkInsert", () =>
...
+                warnings: expect.arrayContaining([
+                  expect.objectContaining({
+                    typeCode:
+                      OfferingValidationWarnings.InvalidFundedStudyPeriodLength,
+                    message:
+                      "The funded study amount of days is ineligible for StudentAid BC funding. Your offering must be at least 12 weeks of study or longer to be eligible.",
+                  }),
+                  expect.objectContaining({
+                    typeCode:
+                      OfferingValidationWarnings.ProgramOfferingDeliveryMismatch,
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.bulkInsert.e2e-spec.ts:248]: We should validate if the array is containing `InvalidFundedStudyPeriodLength`

File: sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.bulkInsert.e2e-spec.ts [hunk 21]
```diff
@@ -221,35 +221,38 @@ describe("EducationProgramOfferingInstitutionsController(e2e)-bulkInsert", () =>
...
+                endDate: "2023-10-14",
+                offeringStatus: OfferingStatus.CreationPending,
+                errors: [],
+                infos: [],
+                warnings: expect.arrayContaining([
+                  expect.objectContaining({
+                    typeCode:
+                      OfferingValidationWarnings.InvalidFundedStudyPeriodLength,
+                    message:
+                      "The funded study amount of days is ineligible for StudentAid BC funding. Your offering must be at least 12 weeks of study or longer to be eligible.",
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.bulkInsert.e2e-spec.ts:244]: I made a comment previously regarding the message having `funded study amount of days`

Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.bulkInsert.e2e-spec.ts:244]: I made a comment previously regarding the message having `funded study amount of days`

File: sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.bulkInsert.e2e-spec.ts [hunk 22]
```diff
@@ -467,6 +470,126 @@ describe("EducationProgramOfferingInstitutionsController(e2e)-bulkInsert", () =>
...
+    async () => {
+      // Arrange
+      // Second location code in the multiple CSV.
+      const csvLocationCodeSEYK = "SEYK";
+      // SABC code in the second row of the multiple CSV.
+      const csvProgramSABCCodeSBC4 = "SBC4";
+
+      // Creating an institution location with same location code as that of the
+      // second row in the CSV file.
+      const collegeFLocationSEYk = createFakeInstitutionLocation(
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.bulkInsert.e2e-spec.ts:486]: minor.  collegeFLocationSEYK ?

File: sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.bulkInsert.e2e-spec.ts [hunk 23]
```diff
@@ -467,6 +470,126 @@ describe("EducationProgramOfferingInstitutionsController(e2e)-bulkInsert", () =>
...
+        {
+          institution: collegeF,
+        },
+        {
+          initialValue: {
+            institutionCode: csvLocationCodeSEYK,
+          } as Partial<InstitutionLocation>,
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.bulkInsert.e2e-spec.ts:493]: The cast is not required.

File: sources/packages/backend/apps/api/src/services/education-program-offering/custom-validators/has-valid-offering-period-for-funded-days.ts [hunk 24]
```diff
@@ -5,78 +5,75 @@ import {
...
-import { OFFERING_STUDY_PERIOD_MAX_DAYS } from "../../../utilities";
 import { StudyBreak } from "../education-program-offering-validation.models";
 import { OfferingCalculationValidationBaseConstraint } from "./offering-calculation-validation-base-constraint";
 
 /**
  * For an offering that contains study breaks, execute the calculation
- * of the funded study period to validate if it matches the mix and max
- * allowed study period amount of days.
+ * of the funded study period to validate if it matches the minimum
+ * allowed study period amount of weeks.
```
Comment (sh16011993) [inline sources/packages/backend/apps/api/src/services/education-program-offering/custom-validators/has-valid-offering-period-for-funded-days.ts:14]: Minor: `funded weeks.`

File: sources/packages/backend/apps/api/src/services/education-program-offering/custom-validators/has-valid-offering-period-for-funded-days.ts [hunk 25]
```diff
@@ -5,78 +5,75 @@ import {
...
       studyBreaks,
       args,
     );
     return (
-      calculatedStudyBreaksAndWeeks.fundedStudyPeriodDays >=
-        offeringMinDaysAllowedValue &&
-      calculatedStudyBreaksAndWeeks.fundedStudyPeriodDays <=
-        OFFERING_STUDY_PERIOD_MAX_DAYS
+      calculatedStudyBreaksAndWeeks.totalFundedWeeks >=
+      offeringMinWeeksAllowedValue
```
Comment (sh16011993) [inline sources/packages/backend/apps/api/src/services/education-program-offering/custom-validators/has-valid-offering-period-for-funded-days.ts:29]: Minor: `offeringMinFundedWeeksAllowedValue`

File: sources/packages/backend/apps/api/src/services/education-program-offering/custom-validators/has-valid-offering-period-for-funded-days.ts [hunk 26]
```diff
@@ -5,78 +5,75 @@ import {
...
+ * @param offeringMinWeeksAllowed study period minimum length in number of weeks.
  * @param validationOptions validations options.
  * @returns true if the study period is valid, otherwise, false.
  */
-export function HasValidOfferingPeriodForFundedDays(
+export function HasValidOfferingPeriodForFundedWeeks(
   startPeriodProperty: (targetObject: unknown) => Date | string,
   endPeriodProperty: (targetObject: unknown) => Date | string,
-  offeringMinDaysAllowed: (targetObject: unknown) => number,
+  offeringMinWeeksAllowed: (targetObject: unknown) => number,
```
Comment (sh16011993) [inline sources/packages/backend/apps/api/src/services/education-program-offering/custom-validators/has-valid-offering-period-for-funded-days.ts:62]: Suggested: `offeringMinFundedWeeksAllowed`

Likewise, please update the comments.

File: sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.bulkInsert.e2e-spec.ts [hunk 27]
```diff
@@ -467,6 +470,126 @@ describe("EducationProgramOfferingInstitutionsController(e2e)-bulkInsert", () =>
...
+        collegeFLocationSEYk,
+      );
+
+      const multipleOfferingFilePath = path.join(
+        __dirname,
+        "bulk-insert/multiple-upload-with-total-funded-weeks-less-than-minimum-allowed-weeks.csv",
+      );
+
+      let responseOfferingSBC1: EducationProgram,
+        responseOfferingSBC2: EducationProgram;
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.bulkInsert.e2e-spec.ts:523]: I know that this has been used from existing tests, but existing ones are not accurate. The offering response from API is `PrimaryIdentifierAPIOutDTO` and more importantly not a program.  Please have it typed the following way so that it will be appropriate. 

image

File: sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.bulkInsert.e2e-spec.ts [hunk 28]
```diff
@@ -467,6 +470,126 @@ describe("EducationProgramOfferingInstitutionsController(e2e)-bulkInsert", () =>
...
+          },
+          order: {
+            name: "ASC",
+          },
+        });
+
+      // The total funded weeks for the programs 'SBC4' in the CSV are less than the minimum allowed weeks,
+      // offering status with 'Creation pending' will be created when inserted.
+      expect(offeringSBC1).toEqual({
+        name: "Test program 41",
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.bulkInsert.e2e-spec.ts:558]: Please use the text `Test offering 41` instead of using `Test program 41`

File: sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.validateOffering.e2e-spec.ts (hunk 14 above)
Comment (dheepak-aot) [reply sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.validateOffering.e2e-spec.ts:204]: I meant `for a full-time education program` should be `for a full-time education program offering`

File: sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.validateOffering.e2e-spec.ts [hunk 29]
```diff
@@ -231,6 +201,543 @@ describe("EducationProgramOfferingInstitutionsController(e2e)-validateOffering",
...
+        studyPeriodBreakdown: {
+          fundedStudyPeriodDays: 229,
+          totalDays: 229,
+          totalFundedWeeks: 33,
+          unfundedStudyPeriodDays: 0,
+        },
+      });
+  });
+
+  it("Should validate an offering with info when valid data with study breaks is passed in for a full-time education program.", async () => {
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.validateOffering.e2e-spec.ts:299]: In description for all E2E tests which are part of this PR, please update `full-time education program` to `full-time education program offering` and like wise for offering.

File: sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.createOffering.e2e-spec.ts [hunk 30]
```diff
@@ -405,6 +405,77 @@ describe("EducationProgramOfferingInstitutionsController(e2e)-createOffering", (
     },
   );
 
+  it(
+    `Should create a new offering with offering status ${OfferingStatus.CreationPending} and warning when data with total funded ` +
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.createOffering.e2e-spec.ts:409]: Please mention the offering intensity in the description.

**Comment by dheepak-aot** [comment]
Thanks for doing the changes @lewischen-aot. Added few more comments. Mostly on E2E area.

File: sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.validateOffering.e2e-spec.ts (hunk 15 above)
Comment (dheepak-aot) [reply sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.validateOffering.e2e-spec.ts]: Please disregard this comment.

**Review by dheepak-aot — APPROVED** [review]
Great Job. Looks good. Thanks for being persistent with the effort required to get the requirements for the messages. 

Not a blocker and if possible, make the branch updated from latest from main so that E2E tests will not fail anymore.

**Review by sh16011993 — APPROVED** [review]
Nice work @lewischen-aot 👍.  Thanks for making the changes.
